aqt/helper.py

```diff
@@ -103,13 +105,30 @@ def downloadBinaryFile(url: str, out: str, hash_algo: str, exp: str, timeout):
                         hash.update(chunk)
                     fd.flush()
             except Exception as e:
-                raise ArchiveDownloadError(f"Download error: {e}") from e
+                raise ArchiveDownloadError(f"Download of {filename} has error: {e}") from e
             if exp is not None and hash.digest() != exp:
-                raise ArchiveDownloadError(
-                    f"Download file is corrupted! Detect checksum error.\nExpected {exp}, Actual {hash.digest()}"
+                raise ArchiveChecksumError(
+                    f"Downloaded file {filename} is corrupted! Detect checksum error.\n"
+                    f"Expected {exp}, Actual {hash.digest()}"
```

Comment:
I found that it was useful to report the filename in the error message. This isn't necessary, but every time I got the checksum error, I wanted to know which file caused it.

Reply:
Is this filename a filename, a relative path or an absolute path?

Reply:
Just the filename. `Path(url).name` is just a really quick way to extract the filename, without the preceding path. We could print the whole url, but all I really wanted to see is the filename.

Reply:
> Just the filename. `Path(url).name` is just a really quick way to extract the filename, without the preceding path.

OK, I didn't read the code before asking that dumb question.

> We could print the whole url, but all I really wanted to see is the filename.

I found that the whole url has already been printed by [`logger.debug("Download URL: {}".format(url))` at installer.py#L899](https://github.com/miurahr/aqtinstall/blob/4eed29fdf1a8c38af73e3fa547eac8fdea327271/aqt/installer.py#L899). However, aqt downloads files in parallel, so the output may be scattered (will it?). In my opinion, it may be more helpful to print the whole URL again, and print the destination path `out` as well. If so, due to the complete URL and path, you would be able to find the local file and manually re-download it from the server more quickly and easily.

The output could be this:

```
...
Downloaded file ${filename} is corrupted! Detect checksum error.
${exp}:           ${url}
${hash.digest()}: ${out}
```

Note that I want to align the url and the path. The length of hash strings are fixed so it can be achieved.

Reply:
> I'm not sure how practical it is to manually re-download the file; when this error occurs, there could be several other files that have not even started downloading yet.

I mean, if we print the URL, developers or experienced users will be able to directly copy-paste it to their browser or HTTP downloader to manually download the file, then do whatever they like.

> How about this for the error message:
> 
> ```
> ...
> Downloaded file ${filename} is corrupted! Detect checksum error.
> Expect ${exp.hex()}:           ${url}
> Actual ${hash.digest().hex()}: ${out}
> ```
> 
> I like having the "Expect" and "Actual" labels there just to clarify which is which.

I agree.

> When `installer()` calls this, it just passes the name of the archive file into `out` without a path, so the file downloads to the working directory. Since `out` has no path info attached to it, I can't tell what it's supposed to be.

OK, I understand it. Those 7z archives contain the proper structure of files and directories, so we can download them to the current working directory (any writable locations are OK) and then extract the content to the specific destination (CWD or `outputdir` ).

How about `str(Path.cwd() / out)` ?

> Both `exp` and `hash.digest()` are "bytes" objects, and if you print them as-is, they look like "\x01\x\23\x45\x67...". The `.hex()` would turn the same value into "01234567". It's half as long, and I think it's easier to read.

Yes, it should definitely be a string. I forgot what the `hash.digest()` function returns.

Reply:
> How about `str(Path.cwd() / out)`?

Yes, that would be an improvement.

I am a little unclear on why we are downloading to the working directory in the first place. If the user happens to have files in the working directory with the same names as the archives downloaded, these files will be overwritten and then deleted. Additionally, if `aqt` fails before finishing this process, the working directory will be littered with `.7z` files that may or may not be partially downloaded. 

I think it might be more respectful of the user's disk space to use a temporary directory to download these files. A temp directory would delete itself automatically when the program ends, whether an error occurs of not. However, this would make it more difficult to attempt to recover an unfinished installation.

Reply:
> ...
> 
> I think it might be more respectful of the user's disk space to use a temporary directory to download these files. A temp directory would delete itself automatically when the program ends, whether an error occurs of not.

I agree. I think downloading archives to `%TMP%` by default makes sense.

This is worth filing a new issue.

> However, this would make it more difficult to attempt to recover an unfinished installation.

Add a new option (parameter) to allow the user to specify the download location? Combine with the `-k, --keep` option (parameter)?

Also, new issue please.

Reply:
> > ...
> > I think it might be more respectful of the user's disk space to use a temporary directory to download these files. A temp directory would delete itself automatically when the program ends, whether an error occurs of not.
> 
> I agree. I think downloading archives to `%TMP%` by default makes sense.
> 
> This is worth filing a new issue.

Done. See #421 .

> > However, this would make it more difficult to attempt to recover an unfinished installation.
> 
> Add a new option (parameter) to allow the user to specify the download location? Combine with the `-k, --keep` option (parameter)?
> 
> Also, new issue please.

Discussion #422 . Talking in the discussion seems more suitable.